runtime/addrmap.c

```diff
@@ -0,0 +1,136 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*      KC Sivaramakrishnan, Indian Institute of Technology, Madras       */
+/*                Stephen Dolan, University of Cambridge                  */
+/*                                                                        */
+/*   Copyright 2020 Indian Institute of Technology, Madras                */
+/*   Copyright 2020 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include "caml/config.h"
+#include "caml/memory.h"
+#include "caml/addrmap.h"
+
+#define MAX_CHAIN 100
+
+static uintnat pos_initial(struct addrmap* t, value key)
+{
+  uintnat pos = (uintnat)key;
+  pos *= 0xcc9e2d51;
+  pos ^= (pos >> 17);
+
+  CAMLassert(Is_power_of_2(t->size));
+  return pos & (t->size - 1);
+}
+
+static uintnat pos_next(struct addrmap* t, uintnat pos)
+{
+  return (pos + 1) & (t->size - 1);
+}
+
+int caml_addrmap_contains(struct addrmap* t, value key)
+{
+  uintnat pos, i;
+
+  CAMLassert(Is_block(key));
+  if (!t->entries) return 0;
+
+  for (i = 0, pos = pos_initial(t, key);
+       i < MAX_CHAIN;
+       i++, pos = pos_next(t, pos)) {
+    if (t->entries[pos].key == ADDRMAP_INVALID_KEY) break;
+    if (t->entries[pos].key == key) return 1;
+  }
+  return 0;
+}
+
+value caml_addrmap_lookup(struct addrmap* t, value key)
+{
+  uintnat pos;
+
+  CAMLassert(Is_block(key));
+  CAMLassert(t->entries);
+
+  for (pos = pos_initial(t, key); ; pos = pos_next(t, pos)) {
+    CAMLassert(t->entries[pos].key != ADDRMAP_INVALID_KEY);
+    if (t->entries[pos].key == key)
+      return t->entries[pos].value;
+  }
+}
+
+static void addrmap_alloc(struct addrmap* t, uintnat sz)
+{
+  uintnat i;
+  CAMLassert(sz > 0 && (sz & (sz - 1)) == 0); /* sz must be a power of 2 */
+  t->entries = caml_stat_alloc(sizeof(struct addrmap_entry) * sz);
+  t->size = sz;
+  for (i = 0; i < sz; i++) {
+    t->entries[i].key = ADDRMAP_INVALID_KEY;
+    t->entries[i].value = ADDRMAP_NOT_PRESENT;
+  }
+}
+
+void caml_addrmap_clear(struct addrmap* t) {
+  caml_stat_free(t->entries);
+  t->entries = 0;
+  t->size = 0;
+}
+
+
+
```

Comment:
These can be trimmed!

Reply:
Updated in 4e084b5721eabeea8d75214bf811d9e0a651b1e5.`